File: 백준/Silver/2178. 미로 탐색/미로 탐색.java
import java.io.*;
import java.util.*;

public class Main {
    public static int n, m;
    public static int[][] arr;
    public static int[][] visited;
    public static Deque<int[]> deque = new ArrayDeque<>();
    public static int[][] ord = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    public static void main(String[] args) throws Exception {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(reader.readLine());
        n = Integer.parseInt(st.nextToken()); m = Integer.parseInt(st.nextToken());
        arr = new int[n][m];
        visited = new int[n][m];
        for (int i = 0; i < n; i++)
            arr[i] = Arrays.stream(reader.readLine().split("")).mapToInt(Integer::parseInt).toArray();
        deque.offer(new int[]{0, 0});
        bfs();
    }

    private static void bfs() {
        while(!deque.isEmpty()) {
            int[] ints = deque.poll();
            if(ints[0] == m - 1 && ints[1] == n - 1) {
                System.out.println(visited[ints[1]][ints[0]] + 1);
                System.exit(0);
            }
            for (int i = 0; i < ord.length; i++) {
                int dx = ints[0] + ord[i][0], dy = ints[1] + ord[i][1];
                if (dx < 0 || dx >= m || dy < 0 || dy >= n)
                    continue;
                else if (visited[dy][dx] == 0 && arr[dy][dx] == 1) {
                    deque.offer(new int[]{dx, dy});
                    visited[dy][dx] = visited[ints[1]][ints[0]] + 1;
                }
            }
        }
    }
}

File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
from collections import deque

n, m = map(int, input().split())
arr = []
for i in range(n): arr.append(input())

directions = [[-1, 0], [0, 1], [1, 0], [0, -1]]
visited = [[False] * m for _ in range(n)]

lst = deque(); lst.append([0, 0, 1]); visited[0][0] = True
while lst:
	cur = lst.popleft()
	if cur[0] + 1 == n and cur[1] + 1 == m:	
		print(cur[2]); break
	for direction in directions:
		y = cur[0] + direction[0]; x = cur[1] + direction[1]
		if y >= n or x >= m or y < 0 or x < 0 or arr[y][x] == '0' or visited[y][x]: continue
		lst.append([y, x, cur[2] + 1])
		visited[y][x] = True